sql: group the optimization sources together, separate planner.analyzeExpr()

knz opened the pull request:
First commit from #20117. 2nd commit is the sub-set of #20096 that is hopefully consensual (the file names don't become longer).

The bare-bones analyzeExpr() is specifically related to index selection,
whereas (*planner).analyzeExpr() is used throughout logical planning for
every SQL scalar expression. Separate them:

- `analyzeExpr()` -> `decomposeExpr()` for the algorithm used in index selection, file
  renamed from `analyze.go` to `opt_decompose.go`.
- `(*planner).analyzeExpr()` remains unchanged but moves to a new file
  `analyze_expr.go`.

The optimization files are then grouped to sort together in file listings.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20169)
<!-- Reviewable:end -->

knz committed:
sql: group the optimization sources together, separate planner.analyzeExpr()

The bare-bones analyzeExpr() is specifically related to index selection,
whereas (*planner).analyzeExpr() is used throughout logical planning for
every SQL scalar expression. Separate them:

- `decomposeExpr()` for the algorithm used in index selection, file
  renamed from `analyze.go` to `opt_decompose.go`.
- `(*planner).analyzeExpr()` remains unchanged but moves to a new file
  `analyze_expr.go`.

The optimization files are then grouped to sort together in file listings.

justinj reviewed (approved):
LGTM

knz commented:
TFYR!
